src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,403 @@
+            }
+
+            // Check if the Mount flag exists; meaning a whole directory has been added under the root, which would require a rescan to rebuild an accurate snapshot.
+            if (shouldRescan == false)
+            {
+                shouldRescan = ((flags & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagMount) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagMount);
+            }
+
+            // Check if the Unmount flag exists; meaning a whole directory has been removed from under the root, which would require a rescan to rebuild an accurate snapshot
+            if (shouldRescan == false)
+            {
+                shouldRescan = ((flags & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagUnmount) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagUnmount);
+            }
+
+            return shouldRescan;
+        }
+
+        private bool DoesPathPassNameFilter(string eventPath)
+        {
+            bool doesPathPass = true;
+            
+            // If we shouldn't include subdirectories, check if this path's parent is the watch directory
+            if (_includeSubdirectories == false)
+            {
+                // Check if the parent is the root. If so, then we'll continue processing based on the name.
+                // If it isn't, then this will be set to false and we'll skip the name processing since it's irrelevant.
+                String parent = System.IO.Path.GetDirectoryName(eventPath);
+                doesPathPass = (parent.Equals(_directory, StringComparison.CurrentCultureIgnoreCase));
+            }
+
+            if (doesPathPass)
+            {
+                // Check if the name fits the pattern
+                doesPathPass = MatchPattern(eventPath);
+            }
+
+            return doesPathPass;
+        }
+
+        private long FindRenameChangePairedChange(
+            long currentIndex, 
+            String[] eventPaths, 
+            Interop.CoreServices.FSEventStreamEventFlags[] eventFlags,
+            FSEventStreamEventId[] eventIds)
+        {
+            Debug.Assert((currentIndex + 1 < eventPaths.Length) && (currentIndex + 1 < eventFlags.Length) & (currentIndex + 1 < eventIds.Length));
+
+            // The ID of the corresponding rename
+            long pairedRenameId = long.MinValue;
+
+            // Start at one past the current index and try to find the next Rename item, which should be the old path.
+            for (long i = currentIndex + 1; i < eventPaths.Length; i++)
+            {
+                if ((eventFlags[i] & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed)
+                {
+                    // We found a match, stop looking
+                    pairedRenameId = i;
```

Comment:
Just a style thing, nothing wrong with what you have, but if it were me, i'd just ```return i;``` here and then ```return long.MinValue;``` at the end of the function, rather than having a local that tracks the information, breaking out the loop, etc.